apps/prairielearn/src/tests/groupExam.test.js

```diff
@@ -0,0 +1,401 @@
+
+const helperServer = require('./helperServer');
+const { idsEqual } = require('../lib/id');
+const { TEST_COURSE_PATH } = require('../lib/paths');
+
+let page, elemList;
+const locals = {};
+locals.helperClient = require('./helperClient');
+locals.siteUrl = 'http://localhost:' + config.serverPort;
+locals.baseUrl = locals.siteUrl + '/pl';
+locals.courseInstanceUrl = locals.baseUrl + '/course_instance/1';
+locals.assessmentsUrl = locals.courseInstanceUrl + '/assessments';
+
+const storedConfig = {};
+
+/**
+ * Switches `config` to new user, loads assessment page, and changes local CSRF token
+ * @param {Object} studentUser
+ * @param {string} assessmentUrl
+ * @param {String} authUin
+ * @param {Number} numCsrfTokens
+ */
+const switchUserAndLoadAssessment = async (studentUser, assessmentUrl, authUin, numCsrfTokens) => {
+  // Load config
+  config.authUid = studentUser.uid;
+  config.authName = studentUser.name;
+  config.authUin = authUin;
+  config.userId = studentUser.user_id;
+
+  // Load assessment
+  const res = await fetch(assessmentUrl);
+  assert.isOk(res.ok);
+  const page = await res.text();
+  locals.$ = cheerio.load(page);
+
+  // Check for CSRF tokens
+  elemList = locals.$('form input[name="__csrf_token"]');
+  assert.lengthOf(elemList, numCsrfTokens);
+  assert.nestedProperty(elemList[0], 'attribs.value');
+  locals.__csrf_token = elemList[0].attribs.value;
+  assert.isString(locals.__csrf_token);
+};
+
+const createGroup = async (groupName, csrfToken, assessmentUrl) => {
+  const form = {
+    __action: 'create_group',
+    __csrf_token: csrfToken,
+    groupName: groupName,
+  };
+  const res = await fetch(assessmentUrl, {
+    method: 'POST',
+    body: new URLSearchParams(form),
+  });
+  assert.isOk(res.ok);
+  locals.$ = cheerio.load(await res.text());
+};
+
+/**
+ * Joins group as current user with CSRF token and loads page with cheerio.
+ * @param {String} assessmentUrl
+ * @param {String} joinCode
+ */
+const joinGroup = async (assessmentUrl, joinCode) => {
+  const form = {
+    __action: 'join_group',
+    __csrf_token: locals.__csrf_token,
+    join_code: joinCode,
+  };
+  const res = await fetch(assessmentUrl, {
+    method: 'POST',
+    body: new URLSearchParams(form),
+  });
+  assert.isOk(res.ok);
+  locals.$ = cheerio.load(await res.text());
+};
+
+describe('Group based exam assess control on student side', function () {
+  this.timeout(20000);
+  before('set authenticated user', function (callback) {
+    storedConfig.authUid = config.authUid;
+    storedConfig.authName = config.authName;
+    storedConfig.authUin = config.authUin;
+    callback(null);
+  });
+  before('set up testing server', helperServer.before(TEST_COURSE_PATH));
+  after('shut down testing server', helperServer.after);
+  after('unset authenticated user', function (callback) {
+    Object.assign(config, storedConfig);
+    callback(null);
+  });
+
+  step('1. the database contains a group-based exam assessment', async function () {
```

Comment:
Nit: don't include numbers in test steps; that makes it harder to refactor to add/remove steps in the future.